pkg/util/signal.go

```diff
@@ -7,32 +7,39 @@ import (
 	"sync"
 )
 
-func NewSingleSignalPair() (SignalSender, SignalReceiver) {
-	sigCh := make(chan struct{})
+func NewSingleSignalPair[T any]() (SignalSender[T], SignalReceiver[T]) {
+	sigCh := make(chan T, 1)
 	once := &sync.Once{}
 	closeSigCh := func() { once.Do(func() { close(sigCh) }) }
 
-	return SignalSender{send: closeSigCh}, SignalReceiver{sigCh: sigCh, closeSigCh: closeSigCh}
+	return SignalSender[T]{
+		send: func(data T) {
+			once.Do(func() {
+				sigCh <- data
+				closeSigCh()
```

Comment:
does calling `closeSigCh` here work? I'd assume it wouldn't actually close the channel, because the `Once` is already running the function, but I don't know how reentrancy is handled

Reply:
You're probably correct. That means we need to just `close(sigCh)` instead of calling `closeSigCh`, and that's safe because both `send` and `closeSigCh` are wrapped in the same once - so we can only have one closure (pun intended), and if we are sending, we know the channel isn't closed.